mmv1/third_party/terraform/provider/provider_project_test.go

```diff
@@ -0,0 +1,257 @@
+					// set
+					t.Setenv("GOOGLE_CLOUD_PROJECT", "GOOGLE_CLOUD_PROJECT") //used
+					t.Setenv("GCLOUD_PROJECT", "GOOGLE_CLOUD_PROJECT")
+					t.Setenv("CLOUDSDK_CORE_PROJECT", "CLOUDSDK_CORE_PROJECT")
+				},
+				Config: testAccSdkProvider_projectInEnvsOnly(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.google_provider_config_sdk.default", "project", "GOOGLE_CLOUD_PROJECT"),
+				),
+			},
+			{
+				// GCLOUD_PROJECT is used 3rd if set
+				PreConfig: func() {
+					// unset
+					t.Setenv("GOOGLE_PROJECT", "")
+					t.Setenv("GOOGLE_CLOUD_PROJECT", "")
+					// set
+					t.Setenv("GCLOUD_PROJECT", "GCLOUD_PROJECT") // used
+					t.Setenv("CLOUDSDK_CORE_PROJECT", "CLOUDSDK_CORE_PROJECT")
+				},
+				Config: testAccSdkProvider_projectInEnvsOnly(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.google_provider_config_sdk.default", "project", "GCLOUD_PROJECT"),
+				),
+			},
+			{
+				// CLOUDSDK_CORE_PROJECT is used 4th if set
+				PreConfig: func() {
+					// unset
+					t.Setenv("GOOGLE_PROJECT", "")
+					t.Setenv("GOOGLE_CLOUD_PROJECT", "")
+					t.Setenv("GCLOUD_PROJECT", "")
+					// set
+					t.Setenv("CLOUDSDK_CORE_PROJECT", "CLOUDSDK_CORE_PROJECT")
+				},
+				Config: testAccSdkProvider_projectInEnvsOnly(),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("data.google_provider_config_sdk.default", "project", "CLOUDSDK_CORE_PROJECT"),
+				),
+			},
+		},
+	})
+}
+
+func testAccSdkProvider_project_emptyStringValidation(t *testing.T) {
+	acctest.SkipIfVcr(t) // Test doesn't interact with API
+
+	context := map[string]interface{}{
+		"project": "",
+	}
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccSdkProvider_projectInProviderBlock(context),
+				PlanOnly:    true,
+				ExpectError: regexp.MustCompile("expected a non-empty string"),
+			},
+		},
+	})
+}
+
+func testAccSdkProvider_project_unknownHandling(t *testing.T) {
+
+	project := envvar.GetTestProjectFromEnv()
+	context := map[string]interface{}{
+		"org_id":             envvar.GetTestOrgFromEnv(t),
+		"billing_account_id": envvar.GetTestBillingAccountFromEnv(t),
+	}
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		ExternalProviders: map[string]resource.ExternalProvider{
+			"random": {},
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testAccSdkProvider_projectUnknownHandling(context),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					// Matches ENV instead of the project id output from google_project resource
+					resource.TestCheckResourceAttr("data.google_provider_config_sdk.default", "project", project),
```

Comment:
Sorry, I am lost somewhere. 
Can you help me understand why `data.google_provider_config_sdk.default.project` should be project here?
 
My understanding is that `data.google_provider_config_sdk.default.project` should be the project in the provider `google`.
The project in the provider `google` should be the value of `google_project.project.project_id`, which should be always be `tf-test-${random_id.project_name.hex}`. 


I also don't understand how the project is unknown in the provider `google`.

Reply:
> * If the `project` config value in the `provider` block is unknown the provider will fall back to the known value coming from ENVs
> * If we remove all project-related ENVs then we force the provider to use the value from  `google_project.project.project_id` once it stops being unknown
> 

Thanks for the information. 
What makes the `project` in the `provider` block is unknown is that `project` needs to be calculated ?
`project = google_project.project.project_id`

Reply:
Yes that's right, if the value is dependent on another resource being provisioned (or a data source fetching a value) then the value is unknown initially in the apply process. Here's an[ analogous example from a user](https://github.com/hashicorp/terraform-provider-google/issues/14255#issuecomment-1523928110) where the credentials value might be supplied from a resource and be temporarily unknown for the same reason ([related GH issue](https://github.com/hashicorp/terraform-provider-google/issues/14444)).

I think that unknown value handling might be best tested at the unit test level, given how it can be impacted by things like dependencies, sleeps, whether a user is applying something in a single apply or not etc. The [flakiness of the test case](https://github.com/GoogleCloudPlatform/magic-modules/pull/11607#issuecomment-2341976030) suggests it's not going to be useful to us in future!